assembly/src/assembler/dead_code_elimination.rs

```diff
@@ -0,0 +1,155 @@
+use alloc::{
+    collections::{BTreeMap, BTreeSet},
+    vec::Vec,
+};
+
+use vm_core::mast::{MastForest, MastNode, MastNodeId};
+
+/// Returns a `MastForest` where all nodes that are unreachable from all procedures are removed.
+///
+/// It also returns the map from old node IDs to new node IDs; or `None` if the `MastForest` was
+/// unchanged. Any [`MastNodeId`] used in reference to the old [`MastForest`] should be remapped
+/// using this map.
+pub fn dead_code_elimination(
+    mast_forest: MastForest,
+) -> (MastForest, Option<BTreeMap<MastNodeId, MastNodeId>>) {
```

Comment:
Returning a `BTreeMap` like this is IMO the weakest part of this PR. An alternative is to move this logic in the `MastForestBuilder`; but I think a larger design change is necessary to ensure that previously returned `MastNodeId`s (e.g. the `body_node_id` of a returned `Procedure` from `MastForestBuilder::get_procedure()`) become cannot be used with the new dead code-eliminated `MastForest`.